File: Bachelor/Semester4/Systems_for_Design_and_Implementation/C#/ProiectCShMPP/Server/service/TripService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using log4net;
using ProiectCShMPP.domain;
using ProiectCShMPP.domain.dtos;
using ProiectCShMPP.repository;

namespace ProiectCShMPP.service
{
    public class TripService
    {
        private readonly ITripRepository _repository;
        private static readonly ILog Logger = LogManager.GetLogger(typeof(TripService));

        public TripService(ITripRepository repository)
        {
            _repository = repository;
        }

        /// <summary>
        /// Method for obtaining all the trips in the database
        /// </summary>
        /// <returns>list of all the trips</returns>
        public IList<Trip> GetAll()
        {
            Logger.InfoFormat("Entry Get All Trips");
            var result = _repository.FindAll().ToList();
            Logger.InfoFormat("Exit Get All Trips {0}", result);
            return result;
        }

        /// <summary>
        /// Method for finding trips that are in a particular destination and timestamp
        /// </summary>
        /// <param name="destination">the desired destination of the trip</param>
        /// <param name="from">the beginning of the timestamp</param>
        /// <param name="to">the end of the timestamp</param>
        /// <returns>list of all the trips in destination between from and to</returns>
        public IList<TripDTO> FindTrips(String destination, DateTime from, DateTime to)
        {
            Logger.InfoFormat("Entry Find trips {0} {1} {2}", destination, from, to);
            var result = _repository.FindTripsByDestinationTime(destination, from, to)
                .Select(trip => new TripDTO()
                {
                    DepartureTime = trip.DepartureTime,
                    TransportFirm = trip.TransportFirm,
                    Price = trip.Price,
                    Seats = trip.Seats
                }).ToList();
            Logger.InfoFormat("Exit Find trips {0}", result);
            return result;
        }

        /// <summary>
        /// Method for updating a trip
        /// </summary>
        /// <param name="trip">the trip to be updated</param>
        /// <returns>
        /// - null, if trip was successfully updated
        /// - trip, otherwise
        /// </returns>
        public Trip Update(Trip trip)
        {
            Logger.InfoFormat("Entry Update trip {0}", trip);
            var result = _repository.Update(trip);
            Logger.InfoFormat("Exit Update Trips {0}", result);
            return result;
        }
    }
}

File: Bachelor/Semester4/Systems_for_Design_and_Implementation/Java/Server/src/main/java/service/TripService.java
package service;

import domain.Trip;
import domain.dtos.TripDTO;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import repository.TripRepository;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

public class TripService {
    private final TripRepository repository;
    private final static Logger logger = LogManager.getLogger();

    public TripService(TripRepository repository) {
        this.repository = repository;
    }

    /**
     * Method for obtaining all the trips in the database
     * @return a list of all the trips
     */
    public List<Trip> getAll(){
        logger.traceEntry();
        var result = StreamSupport.stream(repository.findAll().spliterator(), false).collect(Collectors.toList());
        logger.traceExit(result);
        return result;
    }

    /**
     * Method for obtaining all the trips in a given destination and timestamp
     * @param destination: String, the desired destination of the trip
     * @param from: LocalDateTime, the beginning of the timestamp
     * @param to: LocalDateTime, the end of the timestamp
     * @return a list of all the trips in {@param destination} and between {@param from} and {@param to}
     */
    public List<TripDTO> findTrips(String destination, LocalDateTime from, LocalDateTime to){
        logger.traceEntry("Entry find trips {} {} {}", destination, from, to);
        var result = StreamSupport.stream(repository.findTripsByDestinationTime(destination, from, to).spliterator(), false)
                .map(trip -> new TripDTO(trip.getTransportFirm(), trip.getDepartureTime(), trip.getPrice(), trip.getSeats()))
                .collect(Collectors.toList());
        logger.traceExit(result);
        return result;
    }

    /**
     * Method for updating a trip
     * @param trip: Trip, the trip to be updated
     * @return an {@code Optional}
     *            - null, if trip was successfully updated
     *            - trip, otherwise
     */
    public Optional<Trip> update(Trip trip){
        logger.traceEntry("Entry update Trip {}", trip);
        var result = repository.update(trip);
        logger.traceExit(result);
        return result;
    }
}
